[DC-438] testing directory updates and changes

lrwb-aou opened the pull request:
This PR creates substantial testing changes.

* tests are now in either `curation/tests/unit_tests` or `curation/tests/integration_tests`
* long running tests that **create test data** in a cloud or big query environment should be considered integration tests.
* test directory structures match the `curation/data_steward` directory structure.
* unit tests execute on every push.
* integration tests only execute when you specify "all tests" in a commit message or a PR exists for the branch.
* Coverage information is now available.  The existing unit tests exercise 29% of the python LOC.  View the coverage file for your pushed commit at `https://circleci.com/gh/all-of-us/curation/<build_number>#artifacts/containers/0/test_results/unit/html/index.html`
* Test results are available in Test Summary section.  THERE IS A NICE LOOK AT TEST FAILURE REPORTING FOR PR 607.
* combine integration and unit test results for total test coverage determination.  If any tests were run, you should be able to see which type and coverage results at `<base_url>/0/test_results/all/html/index.html`.

* TODO:  differentiate between a PR and a draft PR.  The information is available in the github payload to circleCI, but is not directly available inside CircleCI.  A workaround needs to be addressed until circleci allows access to the payload.

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,90 @@ jobs:
             t=$(git log -1 --pretty=%B)
-            if [[ "$t" == *"fast tests"* ]] ; then export ALL_TESTS="False" ; else export ALL_TESTS="True" ; fi
-            ./test/run_tests.sh
+            if [[ "$t" == *"all tests"* ]] || [[ ! -z "${CIRCLE_PULL_REQUEST}" ]] || [[ ! -z "${CIRCLE_PULL_REQUESTS}" ]];
```
nit: `-n` is probably a more standard way to write this over `! -z`

lrwb-aou replied:
Done!  I don't spend a lot of time in bash.  Thanks.

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,90 @@ jobs:
+            fi
           no_output_timeout: 3000s
+      - run:
+          name: Comine Coverage Results
```
nit: combine

lrwb-aou replied:
Thank you for catching this!

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,90 @@ jobs:
             t=$(git log -1 --pretty=%B)
-            if [[ "$t" == *"fast tests"* ]] ; then export ALL_TESTS="False" ; else export ALL_TESTS="True" ; fi
-            ./test/run_tests.sh
+            if [[ "$t" == *"all tests"* ]] || [[ ! -z "${CIRCLE_PULL_REQUEST}" ]] || [[ ! -z "${CIRCLE_PULL_REQUESTS}" ]];
```
Will this run after the PR is merged to develop as well, or only while there's a pull request open? I think ideally the integration tests would run on develop as well

lrwb-aou replied:
I've created a ticket for creating nightly build process. DC-636 in our backlog.  This concern should be addressed there.

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,90 @@ jobs:
       - run:
           name: Run unit tests
+          working_directory: ~/curation
+          command: |
```
What is the development flow where someone would only trigger this, and not the integration tests?

In my mind, the usual flow would be:
- push feature branch (starts running CI)
- create PR (attaches to existing CI run)

So based on my understanding, this would just prevent tests from running in the scenario where someone pushed up a branch but either didn't create a PR, or made additional pushes before creating a PR.

Is there a reason this person wants unit tests to run?

lrwb-aou replied:
Yes, there is a reason to run unit tests on each commit.  Unit tests should be fast, therefore, inexpensive in time.  If a developer is updating a portion of code, s/he can make breaking changes (particularly in error prone code).  Unit tests (done properly) will help identify breaking changes before getting to the PR phase.  Because some features require more than one commit to complete, having that information sooner rather than later is useful.

lrwb-aou replied:
That hasn't been the workflow because prior to this, all integration and unit tests were written in the same test package (some in the same module) module and required access to BigQuery services.  While I highly advocate for running unit tests before checking in code, I know it isn't done in practice.  This PR should make is possible to run unit tests locally.  But I don't expect current habits to change.

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,90 @@ jobs:
+                echo "Combining integration and unit test results."
+                cp integration/.coverage all/.coverage.integration
+                cp unit/.coverage all/.coverage.unit
+                cd all
```
nit: pushd all ... popd is probably slightly nicer, as it allows arbitrary nesting of directory changes and is less easily interrupted by things that might have happened between the commands

lrwb-aou replied:
Thank you for the tip.

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,90 @@ jobs:
+          name: Comine Coverage Results
+          working_directory: ~/curation/tests/results/coverage
+          command: |
+            mkdir -p all/xml
```
opt: this is somewhat lengthy, consider moving this into a bash script file instead

lrwb-aou replied:
I did move this to a bash script.  I added `set -x` and `set +x` to the script.

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,90 @@ jobs:
+          command: |
+            mkdir -p tests/results/coverage/unit/xml
+            mkdir -p tests/results/coverage/unit/html
+            mkdir -p tests/results/junit/unit
```
What do the curation tests have to do with junit? Am I missing something? Combine step doesn't seem to touch this

lrwb-aou replied:
You are missing the difference in coverage reporting and CircleCI Test Summary reporting.  The CircleCI platform requires junit xml reporting to use the Test Summary feature.  I like having the test summary data consolidated into the single tab at the beginning of the build, especially when I'm looking for the failures.  

No, the combine step doesn't touch the junit files because it is creating a combined coverage file to show us which lines of our source code were tested and which were not.  I run tests as unit and integration separately, so I have to combine them to get the full report.  Coverage does not use the junit results.  It uses the coverage results.

lrwb-aou replied:
Done.

calbach commented on .coveragerc:
```diff
@@ -1,5 +1,36 @@
+data_file = tests/results/coverage/.coverage
+
+# Exclude files whose paths match these patterns
+omit = *tests*
```
nit: would probably make this slightly more specific, I could imagine "tests" as a valid substring of a real Python file in the curation pipeline.

lrwb-aou replied:
Good call.  Updated to `*curation/tests*`.

calbach commented on .coveragerc:
```diff
@@ -1,5 +1,36 @@
-     *unit_test*
-     *data_steward/lib/*
-     *google-cloud-sdk*
+
```
Is this file still needed? Don't the unit/integration versions subsume this?

lrwb-aou replied:
It is still needed.  It is currently the default coverage configuration file.  It is listed as the default in `tests/runner.py`.  This module makes sure a file with the given name exists and if it does not exist, defaults to `curation/.coveragerc`.

calbach commented on data_steward/dev_requirements.txt:
```diff
@@ -1,3 +1,4 @@
-modernize==0.7
\ No newline at end of file
+modernize==0.7
+unittest-xml-reporting
```
version? At least, a major version probably

lrwb-aou replied:
Done.  `unittest-xml-reporting>=2.5.2,==2.*`

calbach commented on data_steward/tools/combine_ehr_rdr.py:
```diff
@@ -78,10 +78,11 @@ def ehr_consent_query():
-                                                   source_value_ehr_consent=combine_consts.SOURCE_VALUE_EHR_CONSENT,
-                                                   concept_id_consent_permission_yes=combine_consts.
-                                                   CONCEPT_ID_CONSENT_PERMISSION_YES)
+    return combine_consts.EHR_CONSENT_QUERY.format(
```
These appear to be straight formatting changes. Are these intentional? Would be ideal to keep style-only changes separate from this PR

lrwb-aou replied:
There is only one small change that is not a formatting change.  This can be reverted.  Hopefully, yapf will deal with this.  I dislike this formatting.

lrwb-aou replied:
Done.

calbach commented on data_steward/validation/sql_wrangle.py:
```diff
@@ -9,22 +9,32 @@
 TRUNCATE_TABLE_PATTERN = re.compile('\s*truncate\s+table\s+([^\s]+)')
 DROP_TABLE_PATTERN = re.compile('\s*drop\s+table\s+([^\s]+)')
-
+COMMENTED_BLOCK_REGEX = re.compile('(?P<before_comment>(^)(.)*)(?P<comment>(\/\*)(.)*(\*\/))(?P<after_comment>(.)*$)', re.DOTALL)
```
Is this a bad merge perhaps? These changes don't look related and seem like they may be from one of Nishanth's recent PRs.

lrwb-aou replied:
This actually happened.  This module was being tested from main_test.py instead of its own sql_wrangle.py.  So the sql_wrangle tests were removed from main_test.py and put into a created sql_wrangle_test.py module.  In doing this, I wrote a unit test that broke the `is_commented_block()` function.  The previous function couldn't handle true block comments.  This passed as an SQL block comment:
```
-- Some text

-- More text
```

This failed:
```
/*
A block comment

Continuing.
*/
```
So I wrote this to allow the unit test to pass.

lrwb-aou commented:
Hi @calbach ,

- Previously, all tests were stored in a single directory named `curation/data_steward/test/unit_test`.  This PR moves all tests to `curation/tests`.  Python implements its own `test` package, so we were overriding a python package with the original name.
  - Now tests are split based on the resources they use and require.  Unit tests that do not create data or use mocked data are in `curation/tests/unit_tests`.  Integration tests that require putting data in the cloud storage container or Big Query are at `curation/tests/integration_tests`.
  - Multiple existing "unit tests" were split into their appropriate counterparts, some going to unit tests and others going to integration tests.
  - In at least one instance (`curation/tests/unit_tests/data_steward/validation/sql_wrangle_test.py`), a unit test class was created from existing tests that were bundled under other test modules.  In other words, the tests for `main.py` were also validating the `sql_wrangle.py` module.  I added a unit test to this one module that would not pass.  So I also updated the module so it could pass.
-  In unit tests, when possible I replaced calls to get dataset ids with mocked dataset ids.  This is something that typically affects the `setUp()` method.  Now, all the unit tests can easily run on your local development environment.

-  The build config was updated to run all unit tests on every commit, instead of all tests on every commit.
-  The build config was updated to allow running integration tests if a dev specifically wants to test something via integration tests before creating a PR or on a PR.
-  The build config was updated to use CircleCI's Test Summary feature.  It highlights when a failure has occurred as well as slowest tests.
-  The build config was updated to report coverage statistics.  Have you used coverage?  If you go the the Test Artifacts section of the build you can navigate down from `<all/unit/integration>/html/index.html`.  This will display an overview report of which modules were tested and how well.  You can click on the file name links to get specific lines that were and were not tested.

I'm sure there's more.  It is a large change.  If you would like to, we can set up a call.

calbach commented on .circleci/config.yml:
```diff
@@ -55,20 +55,48 @@ jobs:
+          name: Combine Coverage Results
+          working_directory: ~/curation
+          command: |
+              chmod +x tests/combine_coverage.sh
```
nit: should be able to chmod a+x locally and commit the change to git, which obviates having this line

lrwb-aou replied:
done.

lrwb-aou replied:
Done

calbach commented on tests/README.md:
````diff
@@ -0,0 +1,32 @@
+$ PYTHONPATH=.:./data_steward:$PYTHONPATH python tests/runner.py --test-path <optional path>  --test_pattern <optional pattern> --coverage-file <required coverage file path> 
+```
+
+## Directory Structure
````
A few lingering files/directories I see in this PR which look to be misplaced:
- https://github.com/all-of-us/curation/tree/lb/testing-changes/data_steward/test/unit_test/data_steward/cdr_cleaner
- https://github.com/all-of-us/curation/tree/lb/testing-changes/tests/unit_tests/data_steward/data_steward/curation_logging

lrwb-aou replied:
The files at the old location have been added and merged since creating this PR, so the developers put them at the "old location relative to this PR".  I rebased because of merge conflicts and got these.  I'll have to look and see if they are really unit tests or integration tests.  

There's been something weird going on with the `curation_logging` package.  It also messes up when I rebase.  I'll fix it again and it shouldn't be a problem once merged.

lrwb-aou replied:
Had to merge some more test changes in just now.  Rebuilding.

lrwb-aou commented on data_steward/validation/sql_wrangle.py:
```diff
@@ -71,13 +81,38 @@ def temp_repl(m):
+
+    query_list = query_string.split()
+    insert_query = False
+    if query_list[0].lower() in ['insert']:
```
yes.  done.

lrwb-aou commented:
Thanks for checking.  I missed a package initialization file.  Removed.  Confirmed the directory is now gone.
